JordiPolo/lorekeeper
# Rescue errors and ensure fields are reset in JSONLogger

### Changes in this PR
- add `ensure` to `with_extra_fields` method to ensure fields are reset
- fallback to `:object` mode in Oj in case of non representable data
- bugfix in `SimpleLogger` for [non-String messages](https://mdsol.slack.com/archives/C8JH3MH17/p1660852362881679)

### Background
`ensure` keyword change is actually [John's bugfix](https://github.com/JordiPolo/lorekeeper/pull/28), but it was decided to merge both issues in one PR.
The non-representable data in JSON is also [an issue](https://github.com/JordiPolo/lorekeeper/issues/29) introduced by John.

By default `Oj` uses  the`:object` mode which unlike `:compat` supports serialization of almost all classes.
[Here's a table](https://github.com/stereobooster/ruby-json-benchmark/blob/master/test_report.txt) of what is supported in `Oj.dump`  👀 

Also, `:object` [mode doesn't support ](https://github.com/ohler55/oj/blob/develop/pages/Modes.md#options-matrix)`:cache_str` nor `:cache_keys` so no point of passing them.

Not sure if I understood John correctly, but he suggested to log the serialization exception message if everything fails.
Hope my implementation makes sense 🤷‍♀️ 

### How to test

There are specs for the `Oj` fallback change.
You are also always welcome to use [the demo sinatra app](https://github.com/miniengineer/sinatra_with_rack_common_logger_demo_app) just to make sure these changes don't break stuff.
Don't forget to link this branch in the `Gemfile` like so

```
# Gemfile

gem 'lorekeeper', git: 'https://github.com/JordiPolo/lorekeeper', branch: 'feature/rescue-json-logger-errors'
```
